### ISSUE TITLE
Restructure documentation of Behat

### ISSUE THREAD

**Issue by legovaer** [Behat/docs#80]
Hi guys,

I've been thinking over and over again on the structure of the documentation for Behat 3.X. I've noticed that the structure of the documentation of 2.X is rather a bit "weird". Please do not take this as any offense, I'm trying to help here! ;-)

I've been checking a lot of different documentation spaces on readthedocs.org and I found that the documentation of phpspec looks very structured and is easy to read as well (http://www.phpspec.net/en/latest/manual/introduction.html). What do you think about re-structuring the documentation of Behat 3.X? 

Again; this is a suggestion in order to improve the documentation and I'm willing to work on this.

I'd like to have your thoughts on this one.

**Comment by stof** [comment]
The doc may indeed benefit of a reorganization (and of adding lot of content in it as many things are not documented). Neither @everzet nor me have spent time working on the Behat 3 documentation, so any help is definetely welcome.

regarding a new structure of the doc, I suggest you to propose the new structure in a discussion comment here first, and working on applying this new structure only once an agreement is reached to avoid doing work too early and then having to redo it differently.

**Comment by legovaer** [comment]
FYI, I'm currently working on a proposal for the new structure. As soon as I'm finished with this document, I'll share it in this issue.

**Comment by legovaer** [comment]
Here is a first draft of a suggestion. Be advised that there is no specific displaying order. This still needs to be done.

```
1   Introduction
1.1     Behaviour Driven Development
1.2     Why use Behat?
1.3     Features
1.4     Support / Questions
1.5     License

2   Quick Start
2.1     Example
2.2     Installation
2.3     Development
2.3.1           Implementing the Feature
2.3.2           Automating Steps
2.3.3           Defining Steps
2.3.4           Executing Behat

3   User Guide
3.1     Feature Context
3.1.1           Context Traits
3.1.2           Context Parameters
3.1.3           Contexts Lifetime
3.1.4           Context Class Requirements
3.1.5           Using Scopes
3.1.6           Multiple Contexts
3.1.6.1         Gathering Contexts
3.1.7           Hooks
3.1.7.1         Tagged Hooks
3.1.7.2         Scenario Hooks
3.1.7.3         Feature Hooks
3.1.7.4         Step Hooks
3.1.8           Step Definitions
3.1.8.1         Step argument transformation
3.1.8.1.1               Table Transformation
3.1.8.2         Definition Snippets
3.1.8.3         Arguments
3.1.8.4         Definition Keywords
3.2     Defining a Feature
3.2.1           Gherkin Language
3.2.2           Multilingual Support
3.2.3           Tags
3.2.4           Defining a Scenario
3.2.4.1         Tags
3.2.4.2         Pystrings
3.2.4.3         Tables
3.2.4.4         Multiline Arguments
3.2.4.5         Scenario Background
3.2.4.6         Scenario Outline
3.2.4.7         Steps
3.2.4.7.1               Then
3.2.4.7.2               And, But
3.2.4.7.3               Given
3.2.4.7.4               When
3.3     Command Line Tool
3.3.1           Identify Tests
3.3.1.1         By Suite
3.3.1.2         By Tags
3.3.1.3         By Path
3.3.1.4         By Language
3.3.1.5         By Name
3.3.1.6         By Role
3.3.1.7         By Profile
3.3.2           Formatting
3.3.2.1         Define Formatter
3.3.2.2         Specify Output
3.3.2.3         Format Settings
3.3.3           Development Support
3.3.3.1         Append Snippets
3.3.3.2         Dry Running
3.3.3.3         Stop on Failure
3.3.3.4         Rerun
3.3.4           Informative Options
3.3.4.1         Print Definitions
3.3.4.2         Verbose
3.3.4.3         Help
3.3.4.4         Version
3.3.4.5         Config Reference
3.3.4.6         Output Story Syntax
3.3.5           Displaying Colours
3.3.6           Strict
3.3.7           Specifying Configuration file
3.4     Configuration
3.4.1           Paths
3.4.2           Filters
3.4.3           Suites
3.4.3.1         Contexts
3.4.3.2         Formatters
3.4.4           Extensions
3.4.5           Profiles
3.4.6           Importing
3.4.7           Custom Autoloading
3.4.8           Tags
3.5     Extending Behat
3.5.1           Custom Extensions
3.5.1.1         Exceptions
3.5.2           Custom Formatters
3.5.3           Popular Extensions
3.5.3.1         Mink Extension
3.5.3.2         Web Api Extension
3.5.3.3         Drupal Extension
3.5.3.4         Symfony2 Extension

4   Cookbooks
4.1     Configuring Behat for testing Selenium
4.2     Integrating Symfony2 with Behat
4.3     Developing Web Applications with Behat and Mink
4.4     Migrating from Behat 2.x to 3.0
4.5     Using Spin Functions for Slow Tests
4.6     Using the Symfony2 Profiler with Behat and Symfony2Extension
4.7     Intercepting the redirection with Behat and Mink

5   Useful Resources
5.1     Integrate Behat with PHPStorm
5.2     Behat vs Cucumber
5.3     Gherkin API
5.4     Behat API
5.5     Cheat Sheet
```

![behat3-doc](https://cloud.githubusercontent.com/assets/5813212/8454816/0cf77de8-1ff8-11e5-8054-5e62c3c9c951.png)

**Comment by stof** [comment]
Extensions should not be in the Behat doc itself. Each extension should provide its own documentation, which would then follow the versionning of the extension itself, and can be maintained by the extension developers.
Note that once the new behat.org website finally get ready (@everzet any news about when you plan to finish it?), extension docs will appear on it too.

**Comment by legovaer** [comment]
I was not planning to add detailed documentation about Extensions to the docs. The section covering extensions (3.5.3) will only be documented for making the user understand how Behat can be extended. (http://behat.readthedocs.org/en/v2.5/cookbook/behat_and_mink.html). That section will contain a list of extensions and links to their documentation pages.

**Comment by stof** [comment]
Well, your section `3.5.3` is listing some extensions, which should not be part of the Behat documentation itself.

And `Gathering Contexts` should not be in the `Feature context` chapter IMO. This should stay a cookbook. We don't want to document it in the main user guide. It is something you won't need in most cases normally (it is useful when migrating from a Behat 2.5 setup where you used subcontexts)

And each user guide is too big IMO. From the experience of the Symfony doc, small chapters are better than big chapters.

And regarding the order of chapters, talking about gherkin should come before talking about feature contexts IMO. The first step when using Behat is to write feature files, not to write contexts.

**Comment by legovaer** [comment]
> Well, your section 3.5.3 is listing some extensions, which should not be part of the Behat documentation itself.

Fine. Will remove it.

> And Gathering Contexts should not be in the Feature context chapter IMO. This should stay a cookbook. We don't want to document it in the main user guide. It is something you won't need in most cases normally (it is useful when migrating from a Behat 2.5 setup where you used sub contexts)

Not 100% sure if we should not mention it there. As soon as you start using Mink for example, there are a lot of use cases where you have to call that context. Here's an example:

``` php
/**
 * @Given /^I have forgotten my password$/
 */
public function assertForgottenPassword() {
  ... // Some actions here

  // $this->minkContext = \Behat\MinkExtension\Context\MinkContext
  $this->minkContext->assertClick('Forgotten your password?');
}
```

> And each user guide is too big IMO. From the experience of the Symfony doc, small chapters are better than big chapters.

What are you suggesting here? The section 'User Guide' can be compared with `The Book` of Symfony2: A lot of small chapters that are easy to find. Behat is not a framework that can be documented in two pages. There is a need of a user guide.

> And regarding the order of chapters, talking about gherkin should come before talking about feature contexts IMO. The first step when using Behat is to write feature files, not to write contexts.

100% right. But the order of the content is currently not in scope for the pasted suggestion. I suggest to first focus on all the content and then start ordering.

**Comment by stof** [comment]
@legovaer **none** of the Mink assertion are implemented in the MinkContext itself. they are all in the WebAssert class of Mink itself. This means that you should just make your context MinkAware to have access to Mink and use the Mink API rather than trying to gather contexts to access the MinkContext class (when writing your features in a business oriented way, the MinkContext class should not even be registered in your suite as the builtin Mink step are talking the browser language, not the language of your business)

> What are you suggesting here? The section 'User Guide' can be compared with The Book of Symfony2: A lot of small chapters that are easy to find. Behat is not a framework that can be documented in two pages. There is a need of a user guide.

I agree with you that a user guide is needed. what I'm saying is that it is impossible to have a single chapter in it with everything you listed under `Feature Context`. It would make it far too big. The comparison with the Symfony book is a good one. Most chapters are much shorter than what you describe it. And the Symfony doc team actively tries to maintain them short, moving some doc out of the book regularly, to split it in new chapters (often in cookbooks describing how to do specific things)

And the suggestions you are making for the sublevel of titles don't all look like chapters (and none of them look like chapter titles, even though some of them may be logical with a better wording), so we cannot consider that chapters are one level deeper in your proposal.

> 100% right. But the order of the content is currently not in scope for the pasted suggestion. I suggest to first focus on all the content and then start ordering.

I think it should actually be the opposite, because the order in which we put chapters will impact the way content is written a lot (we generally expect people to know what is in the previous section, but not what is in the next one)

**Comment by legovaer** [comment]
Apparently I will have to do some research on the implementation of extensions in Behat 3.0. I think I've done it all wrong on my current project (never the less, it's working). If what you are saying makes sense, then I can agree that it's not necessary to add that information to the part `User Guide`.

 I will split up the `Feature Context` section and make sure that's easy to understand to see what is a chapter or what is content of a chapter.

Makes sense to me, I'll start sorting this stuff.

**Comment by legovaer** [comment]
@stof here's another go;

```
1   Introduction
1.1     Behaviour Driven Development
1.2     Why use Behat?
1.3     Features
1.4     Example
1.5     Support / Questions
1.6     License
2   Quick Start
2.1     Installation
2.2     Development
2.2.1           Implementing the Feature
2.2.2           Automating Steps
2.2.3           Defining Steps
2.2.4           Executing Behat
3   User Guide
3.1     About Gherkin Language
3.2     Features and Scenario’s (Basic examples)
3.3     Initializing a new Behat project
3.4     Creating features
3.5     Writing scenarios (Steps (Given, When, Then, And & But), Background, Outline, Tables, Multiline Arguments, Pystrings)
3.6     Multilingual support in features
3.7     Identifying features and scenarios
3.7.1           Basic Tagging
3.7.2           Magic Tagging (Using tags in context)
3.8     Feature Contexts
3.8.1           #Context Traits
3.8.2           #Context Parameters
3.8.3           #Contexts Lifetime
3.8.4           #Context Class Requirements
3.8.5           #Using Scopes
3.8.6           #Multiple Contexts
3.8.7           Hooking into the test process
3.8.7.1         Tagged Hooks
3.8.7.2         Scenario Hooks
3.8.7.3         Feature Hooks
3.8.7.4         Step Hooks
3.8.8           Defining Step Definitions
3.8.8.1                 #Step argument transformation (+ Table transformation)
3.8.8.2                 #Definition Snippets
3.8.8.3                 #Arguments
3.8.8.4                 #Definition Keywords
3.9     Command Line Tool
3.9.1           Identifying tests (By: suite, tags, path, language, name, role & profile)
3.9.2           Formatting the Behat output (Define formatter, specify o
3.9.2.1         #Define Formatter
3.9.2.2         #Specify Output
3.9.2.3         #Format Settings
3.9.3           Development support (Append snippets, dry running, stop on failure, rerun)
3.9.4           Informative output (Print definitions, verbose, help, version, config reference, output story syntax)
3.9.5           Displaying Colours
3.9.6           Strict
3.9.7           Specifying configuration file
3.10        Configuration
3.10.1      Feature, Suite and Scenario configuration  (Paths, Filters, Suites, Contexts, Extensions)
3.10.2      #Profiles
3.10.3      #Importing
3.10.4      #Custom Autoloading
3.10.5      #Formatters
3.11        Extending Behat
3.11.1      Developing a custom extension
3.11.1.1            Exceptions (Overview of Behat exceptions that a developer can use)
3.11.2      Developing a custom formatter
4   Cookbooks
4.1     Configuring Behat for testing Selenium
4.2     Integrating Symfony2 with Behat
4.3     Developing Web Applications with Behat and Mink
4.4     Migrating from Behat 2.x to 3.0
4.5     Using Spin Functions for Slow Tests
4.6     Using the Symfony2 Profiler with Behat and Symfony2Extension
4.7     Intercepting the redirection with Behat and Mink
4.8     Gathering contexts when using multiple contexts
5   Useful Resources
5.1     Integrate Behat with PHPStorm
5.2     Behat vs Cucumber
5.3     Gherkin API
5.4     Behat API
5.5     Cheat Sheet
```

**Comment by stof** [comment]
I think `Multilingual support in features` might be moved to a cookbook (I doubt many projects would actually use multiple languages).
and `Identifying features and scenarios` is also more advanced.

Regarding your `Feature Contexts` chapter, I suggest splitting it in multiple chapters. For instance, hooks may have a dedicated chapter.

**Comment by stof** [comment]
@weaverryan if you have some time, it would be great to have your input on this discussion to benefit from your experience btw.

**Comment by weaverryan** [comment]
I think what @legovaer has proposed is very well thought-out indeed - the only thing that stuck out was what stof already mentioned - move "Multilingual" to a cookbook, just because I wouldn't want something that's uncommon having such a "high-up" spot on the docs. Oh, and maybe also have configuration be its own chapter - and put "behat.yml" in the title name (see my point 3 below).

Fwiw, I'm looking at this asking:

1) Is it obvious where I can install?
2) Is there a quick start where I can work for 5 minutes and get some initial success and excitment?
3) Are the user guide titles well-titled and small enough that they'll come up on common google searches (e.g. "Behat FeatureContext", or "behat.yml configuration")
4) Is edge-case stuff pushed into their own articles (instead of taking up space in the middle of important articles).

I think this is well-thought out - I would love to see Behat's docs get love to get it into this format :).

Cheers!

**Comment by legovaer** [comment]
Thanks for that positive feedback @weaverryan. With that being said, here’s my suggested workflow in order to get the Behat documentation updated ASAP:

![suggested-workflow](https://cloud.githubusercontent.com/assets/5813212/8575304/34254774-2593-11e5-8a4c-5fbf8d5e3afc.png)
1. I will fork this repository from the v3.0 branch and will create a new branch: ‘legovaer/docs/fishbone’. This branch will be used in order to develop the fishbone for the new structure of the documentation. As soon as this fishbone is ready, this branch will become the base branch for all further development.
2. Now that the new structure is available, everyone who would like to participate can fork from the branch ‘legovaer/docs/fishbone’ and can start adding content.
3. As soon as all content has been added, a final review will be done by the community (we will have to promote this in order to ‘awake’ the community). For every bug found in the documentation, a new issue will be created in the legovaer/docs repository. The fixing of this bug will be done in a separate branch.
4. When the code review has been finished, the final PR will be created and merged back into Behat/docs

**Remarks**
- My suggestion is to work in phases. During the first phase (MVP phase), we will develop the fishbone and only add chapters that already can be filled with existing content. This allows us to deliver in a quick way a structured guide without too much effort. During the next phase we can start adding content to the fishbone in small sprints.
- In the Diagram I used ‘stof/docs/chapter y’ as an example to indicate that everyone will be able to contribute.
- It might be a good idea to get in touch with some SEO / content expert during the final review phase.

**Comment by legovaer** [comment]
@stof @weaverryan @everzet is the suggested approach described above ok for you guys? I'd like to start working on it! ;-)

**Comment by weaverryan** [comment]
@stof is on holiday - and @everzet might have a thought, but my guess is that he'll mostly be thrilled to have some great and thoughtful help on this issue, no matter what :). +1 from me - I can subscribe to your repo and proof the PR's over there. If the flow isn't working out - we can always change gears.

So, for me, go go go! :)

**Comment by legovaer** [comment]
The new fishbone has been developed and pushed to the [legovaer/docs/Fishbone branch](https://github.com/legovaer/docs/tree/fishbone).

**Updates:**
- Developed a new fishbone structure from scratch
- Temporary removed all existing documentation
- Switched (temporary?) to the default RTF theme because there was no clear documentation structure visible in the `borg` theme.

**Comment by stof** [comment]
@legovaer wouldn't it be possible to migrate documentation little by little in the main repo rather than doing all work in a separate repo and then a single huge PR impossible to review on this repo ? It would also mean that improvements are online as soon as they are ready.

and I was indeed in vacation during 10 days, which is why I haven't been responsive here

**Comment by legovaer** [comment]
@stof I hope you had a great holiday! 

Your comment is a little bit too late. I just finished [migrating all the content](https://github.com/legovaer/docs/tree/mvp) :( 

I can still create small chunks if that might be helpful. 

**Updates:**
- Migrated all existing content inside the new fishbone structure
- Updated the theme in order to have a structured overview of chapters
- Updated fishbone 1.0 after migrating the existing content. Some structure didn't make any sense at all.
- Added the `output/` folder to simplify the reviewing process. This will be removed as soon as the final PR is going to be created.
- Created [ReadTheDocs project](http://behat-docs.readthedocs.org/en/mvp1.0/) in order to simplify the reviewing process.

## Proposed workflow

### Fishbone

The proposed new structure can be found in the [legovaer/docs/fishbone](https://github.com/legovaer/docs/tree/fishbone) branch. As soon as something needs to be changed in the entire structure, this branch needs to be updated. This branch contains _no_ content and is only used for having a clear understanding of the documentation structure

### Adding new content

As soon as the [MVP branch](https://github.com/legovaer/docs/tree/mvp) has been merged into Behat/Docs, we will be able to start adding content to the new structure. We should keep in mind (again) that as soon as we do any structural changes in the documentation, we should update Fishbone.

### When all content is added to Fishbone

We will be able to drop the [legovaer/docs/fishbone](https://github.com/legovaer/docs/tree/fishbone) branch and continue working with the behat/docs branch.

**Comment by stof** [comment]
@legovaer instead of adding the output folder, a solution could be to enable ReadTheDocs on your repo (it is a free tool) to have the doc rendered on each commit in the fishbone branch (exactly as we do in Behat). Instead of reviewing the rendered HTML (which is harder than reviewing the rST source), it would allow to browse the new doc.

I don't understand the goal of the separate fishbone and MVP branch though. Which one is built on top of the other ? Your description tells me that MVP is on top of Fishbone (as it has the content and Fishbone does not), but then you talk about adding content to Fishbone, but about merging MVP in Behat

**Comment by weaverryan** [comment]
Side note - I do think that this is a big enough change, that a piece-by-piece review might not be useful. I was envisioning some HUGE changes, then reviewing sections as they're finished (by actually reading through them).

**Comment by legovaer** [comment]
@stof The fishbone branch contains only headers (documentation structure). Since there is a lot of documentation missing, I want to avoid that we add empty sections to the live documentation. We should always keep Fishbone as the example of the structure.

The MVP branch is forked of Fishbone where I have removed empty sections and added content.

As soon as we have all content available (every single header from Fishbone has been filled), we can get rid of that Fishbone branch.

I'm currently setting up ReadTheDocs.

@weaverryan I agree with you. Splitting up all changes will cost a lot of effort because I changed the URLs as well (improve SEO).

**Comment by stof** [comment]
We should try to keep files at old URLs with links to the new locations of the content though, to avoid breaking existing links to the Behat documentation (actually making the break less an issue).

**Comment by weaverryan** [comment]
Do we have a way to setup 301's?

**Comment by legovaer** [comment]
@stof No, we should setup [user-defined redirects](https://read-the-docs.readthedocs.org/en/latest/user-defined-redirects.html). This will return a 301 redirect which improves your SEO again. It will be a lot of work because there's currently no automated process for that. But if you want, I can help with that as well.

**Comment by legovaer** [comment]
Here are the RTD links:

[MVP](http://behat-docs.readthedocs.org/en/mvp1.0/)

[Fishbone](http://behat-docs.readthedocs.org/en/fishbone/)

**Comment by stof** [comment]
@legovaer I suggest sending a PR to the behat/docs repo as soon as all the existing content is migrated to the new structure, and then continuing the work by doing new PRs to Behat/docs instead of doing all the work in your fork. This will add 2 advantages:
- it will avoid to have changes done in the old doc which would then have to be re-applied in the restructured doc (or be lost). Be aware that this happened since you started as @everzet merged some PRs yesterday
- it will give more chances to have community contributions as progress will happen in the main repo
- it will publish the improved doc sooner (adding new content in the documentation can perfectly be done separately anyway)

If we wait to have all the wished documentation written before merging the improved structure, we risk waiting a long time in case contributors have less time to write the doc (and writing lots of doc takes lots of time)

**Comment by legovaer** [comment]
@stof sorry for the late response but I was on vacation. I agree with your approach. I noticed that there was some content added to the main repository, so I suggest that I merge this added content into mine before creating the PR. 

With regard to the 'fishbone' (the structure of the documentation as we suggest to have it in the future), I think it will be useful to document this somewhere so that all the work of this issue wasn't just for nothing. I see two possible options:
- Adding the fishbone to my PR by creating empty pages, chapters and sections.
- Adding the fishbone as a guide to the README file of the doc repository.

When thinking about SEO, I think the best solution will be the second option. What do you guys think?

**Comment by stof** [comment]
I would rather use the issue tracker for that rather than the readme. We should have an issue per missing chapter/section in the fishbone, which can be closed when someone contributes.

**Comment by legovaer** [comment]
Agreed. I will create the PR. As soon as the PR has been merged, I will create all the issues.
